common/io/src/buffer/nested_checkpoint_reader.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io;
+use std::io::Result;
+
+use super::BufferRead;
+
+// this struct can avoid NestedCheckpointReader<NestedCheckpointReader<R>> in recursive functions
+pub struct NestedCheckpointReader<R: BufferRead> {
+    reader: R,
+    // clear buffer until all checkpoints is popped and the buffer is consumed
+    buffer: Vec<u8>,
+    checkpoints: Vec<usize>,
+
+    checkpointing: bool,
+    pub pos: usize,
+}
+
+impl<R: BufferRead> NestedCheckpointReader<R> {
+    pub fn new(reader: R) -> Self {
+        Self {
+            reader,
+            buffer: Vec::new(),
+            checkpoints: Vec::new(),
+            checkpointing: false,
+            pos: 0,
+        }
+    }
+
+    pub fn push_checkpoint(&mut self) {
+        self.checkpointing = true;
+        self.checkpoints.push(self.pos)
+    }
+
+    // todo(youngsofun): add CheckPointGuard to make it safer.
```

Comment:
Will be great to have a Guard

Reply:
I will do it in the next pr, with some other minor refactors.  let`s fix the bug first.